### DISCUSSION TITLE
linking s3 prefix to iterables in tar_map()

### DISCUSSION THREAD

**Discussion by gofford** [ropensci/tarchetypes#62] [Help] [answered]
When setting up targets with S3 artefacts, the standard setup is something like this:

```
tar_option_set(
    resources = tar_resources(
      aws = tar_resources_aws(
        bucket = BUCKET,
        prefix = S3_PREFIX
      )
    )
  )
```

This necessarily stores all artefacts for targets in the same location on S3. For (very) large target DAGs, which span multiple iterables using `tar_map()`, is there a way of augmenting the `BUCKET/S3_PREFIX` such that the different mapped values are sub-paths within `BUCKET/S3_PREFIX`? i.e., if I'm `tar_map`ping over `values = list('A', 'B', 'C')`, is there a way to configure the artefacts to be stored in `S3_PREFIX/A`, `S3_PREFIX/B` ... etc? 

This is a human usability point rather than a programmatic one. The S3 scrolling and search UI is not great from a usability perspective which makes it hard to find specific artefacts when there are large numbers of them.

The closest I have been able to replicate this is to move the `map` outside of the target factory rather than inside it; i.e.,

```
list('A', 'B', 'C') %>% 
     walk(function(val) target_factory(val))
```

where `tar_resources_aws()` is configured within the factory. However, this breaks the DAG into many parts which is awkward from a global reusability perspective. Is there a way of achieving this within `tar_map()` (or a target factory)?

**Comment by wlandau** [comment, marked as answer]
By design, `tar_map()` only manipulates the commands and dynamic branching patterns of existing target objects. `tar_eval()`, on the other hand, will let you manipulate an arbitrary `tar_target()` call before it is evaluated, including settings like the bucket and prefix: https://books.ropensci.org/targets/static.html#metaprogramming. (Alternatively, you could also loop over existing target objects and modify `$settings$resources$aws$bucket` and `$settings$resources$aws$prefix` for each target object, but technically speaking I cannot make any promises to keep the internal structure of target objects the same from one release to the next.)

**Reply by gofford** [reply]
Do you have any examples of sophisticated use of `tar_eval()` to replicate, e.g., `tar_combine()`-like operations, etc? I would certainly benefit from this as it is a lot less intuitive than native `tar_map()`. Either way, I can definitely see that this is the correct method for this. However, it's a pretty big undertaking to re-write an existing pipeline (with dynamic-within-static branching) so I'll leave it as-is for now and return to it later. Thanks again!

**Reply by gofford** [reply]
Any examples of replicating static multi-branch and dynamic branching with `tar_eval()` would be a great addition to future documentation. Complexity with metaprogramming goes from 0 to light-speed in no time at all.

**Reply by wlandau** [reply]
In this situation, tar_eval() is a direct substitute for tar_map(). Both can return lists of targets, and tar_combine() works the same way on the resulting target lists. The only difference is that tar_eval() acts on the expression you type (e.g. tar_target(…)) rather than the object returned when that expression is evaluated. In other words, usage is not very different, and you have additional flexibility.
